integration/cdi/src/main/java/org/apache/cxf/cdi/JAXRSCdiResourceExtension.java

```diff
@@ -372,7 +379,14 @@ private ClassifiedClasses classes2singletons(final Application application, fina
 
             for (final Bean< ? > bean: serviceBeans) {
                 if (classes.contains(bean.getBeanClass())) {
-                    classified.addResourceProvider(new CdiResourceProvider(beanManager, bean));
+                    final CdiResourceProvider provider = new CdiResourceProvider(beanManager, bean);
+                    // if not a singleton we manage it per request
+                    // if @Singleton the container handles it
+                    // so we only need this case here
+                    if (provider.isSingleton() && Dependent.class == bean.getScope()) {
+                        disposableResourceProviders.add(provider);
```

Comment:
```
synchronized (disposableResourceProviders) {
    disposableResourceProviders.add(provider);
}
```

Reply:
Oh good catch, copied the pattern of disposableCreationalContexts but actually we don't need that for that list which is just used in the context of the extension so thread safe, will clean up